### PR TITLE
Add support for PR reviews preview

### PR DESCRIPTION
**Pull request by stephenc**
@reviewbybees

### CODE REVIEW

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2010, Kohsuke Kawaguchi
```
Comment (KostyaSha) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: really?

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 2]
```diff
@@ -224,6 +228,27 @@ protected void wrapUp(GHPullRequestFileDetail[] page) {
     }
 
     /**
+     * Retrieves all the reviews associated to this pull request.
+     */
```
Comment (oleg-nenashev) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: Nice2have: since

Comment (oleg-nenashev) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: And below

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    protected String getApiRoute() {
+        return owner.getApiRoute()+"/reviews/"+id;
+    }
+
+    /**
+     * Updates the comment.
+     */
+    @Preview
+    @Deprecated
```
Comment (oleg-nenashev) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: Maybe it worth adding dependency on accmod and adding `@Restricted` annotations instead

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 4]
```diff
@@ -224,6 +228,27 @@ protected void wrapUp(GHPullRequestFileDetail[] page) {
...
+    public PagedIterable<GHPullRequestReview> listReviews() {
+        return new PagedIterable<GHPullRequestReview>() {
+            public PagedIterator<GHPullRequestReview> _iterator(int pageSize) {
+                return new PagedIterator<GHPullRequestReview>(root.retrieve()
+                        .withPreview("application/vnd.github.black-cat-preview+json")
```
Comment (henryju) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: What about adding an entry in Previews class?

**Comment by porcelli** [comment]
Hi, I'm interested in this PR, actually I've even enchanced this with support of new Events (Status and PullRequestReview). But to submit my PR, I'd need this merged first.

**Comment by jglick** [comment]
I think only @kohsuke can do that.

**Comment by henryju** [comment]
Fixes #305

**Comment by PauloMigAlmeida** [comment]
I have accidentally created a  pull request for the same purpose (#366) and I must confess that I'm very surprised that it's similar in so many ways. Regardless of that, since this PR has been created before and has already been reviewed by a CloudBees employee, I will close mine in favour or this one.

I will do my review on this one too since we both spent a significant time working on this matter and I would love to see this merged in the short-term as some of the automation process I created at work would really benefit of it.

@kohsuke when you have a spare time, please merge it ;)

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 5]
```diff
@@ -300,4 +353,28 @@ private void fetchIssue() throws IOException {
             fetchedIssueDetails = true;
         }
     }
+
+    private static class DraftReviewComment {
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: It's just my 2 cents opinion but why don't you make this public and add the GH prefix on it too to make it standard with the majority of the other classes in the project.

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 6]
```diff
@@ -259,6 +284,34 @@ protected void wrapUp(GHPullRequestCommitDetail[] page) {
...
+        if (event == null) {
+            event = GHPullRequestReviewState.PENDING;
+        }
+        List<DraftReviewComment> draftComments = new ArrayList<DraftReviewComment>(comments.size());
+        for (GHPullRequestReviewComment c : comments) {
+            draftComments.add(new DraftReviewComment(c.getBody(), c.getPath(), c.getPosition()));
+        }
+        return new Requester(root).method("POST")
+                .with("body", body)
+                //.with("event", event.name())
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: Correct me if I'm wrong but if you aren't sending the event, then all of the reviews will be set to Pending state, right? If that's the case, why you have the event parameter in the first place?

I think we should let send it as it was specified in the method signature (2 cents opinion again)

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 7]
```diff
@@ -259,6 +284,34 @@ protected void wrapUp(GHPullRequestCommitDetail[] page) {
...
+                ._with("comments", draftComments)
+                .withPreview("application/vnd.github.black-cat-preview+json")
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 8]
```diff
@@ -259,6 +284,34 @@ protected void wrapUp(GHPullRequestCommitDetail[] page) {
...
 
+    @Preview
+    @Deprecated
+    public GHPullRequestReview createReview(String body, @CheckForNull GHPullRequestReviewState event,
+                                            GHPullRequestReviewComment... comments)
+            throws IOException {
+        return createReview(body, event, Arrays.asList(comments));
+    }
+
+    @Preview
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 9]
```diff
@@ -259,6 +284,34 @@ protected void wrapUp(GHPullRequestCommitDetail[] page) {
...
+    @Deprecated
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequest.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    protected String getApiRoute() {
+        return owner.getApiRoute()+"/reviews/"+id;
+    }
+
+    /**
+     * Updates the comment.
+     */
+    @Preview
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java (hunk 3 above)
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    @Deprecated
+    public void submit(String body, GHPullRequestReviewState event) throws IOException {
+        new Requester(owner.root).method("POST")
+                .with("body", body)
+                .with("event", event.action())
+                .withPreview("application/vnd.github.black-cat-preview+json")
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                .to(getApiRoute()+"/events",this);
+        this.body = body;
+        this.state = event;
+    }
+
+    /**
+     * Deletes this review.
+     */
+    @Preview
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    @Deprecated
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    public void delete() throws IOException {
+        new Requester(owner.root).method("DELETE")
+                .withPreview("application/vnd.github.black-cat-preview+json")
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+                .to(getApiRoute());
+    }
+
+    /**
+     * Dismisses this review.
+     */
+    @Preview
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    @Deprecated
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    public void dismiss(String message) throws IOException {
+        new Requester(owner.root).method("PUT")
+                .with("message", message)
+                .withPreview("application/vnd.github.black-cat-preview+json")
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+     * Obtains all the review comments associated with this pull request review.
+     */
+    @Preview
+    @Deprecated
+    public PagedIterable<GHPullRequestReviewComment> listReviewComments() throws IOException {
+        return new PagedIterable<GHPullRequestReviewComment>() {
+            public PagedIterator<GHPullRequestReviewComment> _iterator(int pageSize) {
+                return new PagedIterator<GHPullRequestReviewComment>(
+                        owner.root.retrieve()
+                                .withPreview("application/vnd.github.black-cat-preview+json")
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+                .with("message", message)
+                .withPreview("application/vnd.github.black-cat-preview+json")
+                .to(getApiRoute()+"/dismissals");
+        state = GHPullRequestReviewState.DISMISSED;
+    }
+
+    /**
+     * Obtains all the review comments associated with this pull request review.
+     */
+    @Preview
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReview.java [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+    @Deprecated
```
Comment (PauloMigAlmeida) [inline src/main/java/org/kohsuke/github/GHPullRequestReview.java]: This is no long a preview method, so you can remove this line

File: src/main/java/org/kohsuke/github/GHPullRequestReviewState.java [hunk 21]
```diff
@@ -0,0 +1,19 @@
+package org.kohsuke.github;
+
+public enum GHPullRequestReviewState {
+    PENDING(null),
+    APPROVED("APPROVE"),
```
Comment (jaroslaw-bochniak) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewState.java]: Isn't this a typo `APPROVE` -> `APPROVED`?  At least GH Api docs are listing the `APPROVED` state https://developer.github.com/v3/pulls/reviews/#get-a-single-review

**Comment by mattnelson** [comment]
> Hi, I'm interested in this PR, actually I've even enchanced this with support of new Events (Status and PullRequestReview). But to submit my PR, I'd need this merged first.

@porcelli Now that this is merged and released would you be able to issue your PR to add event support for PR reviews?

**Comment by PauloMigAlmeida** [comment]
@mattnelson the new events (Status and PullRequestReview) that @porcelli mentioned were already merged in this PR: https://github.com/kohsuke/github-api/pull/363/files

**Comment by mattnelson** [comment]
@PauloMigAlmeida thanks, I missed that.

I was specifically looking for the data binding in `GHEventPayload` as I was wanting to implement a hook on the PullRequestReview event. If this hasn't been implemented yet, I will look into it. The rest of the parts are already there, so it should be pretty strait forward.

**Comment by PauloMigAlmeida** [comment]
Hi @mattnelson 

If what you want it to implement a hook based on the ```PullRequestReview``` event from ``GHEvent``` class. This is what you can do:
PS.: This is a piece of code extracted from an internal system of ours that does exactly you just described.

```Java
public class GitConfigRepositoryOperation extends AbstractGitHubOperation<Void> {

    private GHRepository repository;
    private final static String webhookUrl;

    static{
        webhookUrl = KMSEnvironmentalVariableDecryptService.getInstance().decryptKey("GitWebhookUrl");
    }

    public GitConfigRepositoryOperation(GHRepository repository) throws IOException {
        super(repository.getFullName());
        this.repository = repository;
    }

    @Override
    public Void performOperation() throws Exception {

        // Configure Webhook
        Map<String, String> configMap = new HashMap<>(2);
        configMap.put("url", webhookUrl);
        configMap.put("content_type", "json");
        this.repository.createHook("web",
                configMap,
                Arrays.asList(
                    GHEvent.PULL_REQUEST,
                    GHEvent.PULL_REQUEST_REVIEW,
                    GHEvent.PULL_REQUEST_REVIEW_COMMENT,
                    GHEvent.PUSH,
                    GHEvent.CREATE
                ), true);

        return null;
    }
```

Hope that it helps.

**Pull request merged**